examples/pipeline/wav2letter.py

```diff
@@ -0,0 +1,1179 @@
+
+lstm_params = {
+    "hidden_size": 800,
+    "num_layers": 5,
+    "batch_first": False,
+    "bidirectional": False,
+    "dropout": 0.,
+}
+
+clip_norm = 0.  # 10.
+
+zero_infinity = False
+
+
+# Text encoding
+
+class Coder:
+    def __init__(self, labels):
+        labels = [l for l in labels]
+        self.length = len(labels)
+        enumerated = list(enumerate(labels))
+        flipped = [(sub[1], sub[0]) for sub in enumerated]
+
+        d1 = collections.OrderedDict(enumerated)
+        d2 = collections.OrderedDict(flipped)
+        self.mapping = {**d1, **d2}
+
+    def encode(self, iterable):
+        if isinstance(iterable, list):
+            return [self.encode(i) for i in iterable]
+        else:
+            return [self.mapping[i] + self.mapping[char_blank] for i in iterable]
+
+    def decode(self, tensor):
+        if isinstance(tensor[0], list):
+            return [self.decode(t) for t in tensor]
+        else:
+            # not idempotent, since clean string
+            x = (self.mapping[i] for i in tensor)
+            x = ''.join(i for i, _ in itertools.groupby(x))
+            x = x.replace(char_blank, "")
+            # x = x.strip()
+            return x
+
+
+coder = Coder(labels)
+encode = coder.encode
+decode = coder.decode
+vocab_size = coder.length
+print("vocab_size", vocab_size, flush=True)
+
+
+# Model
+
+model = Wav2Letter(num_features, vocab_size)
+
+def model_length_function(tensor):
+    return int(tensor.shape[0])//2 + 1
+
+
+# Dataset
+
+
+class IterableMemoryCache:
+
+    def __init__(self, iterable):
+        self.iterable = iterable
+        self._iter = iter(iterable)
+        self._done = False
+        self._values = []
+
+    def __iter__(self):
+        if self._done:
+            return iter(self._values)
+        return itertools.chain(self._values, self._gen_iter())
+
+    def _gen_iter(self):
+        for new_value in self._iter:
+            self._values.append(new_value)
+            yield new_value
+        self._done = True
+
+    def __len__(self):
+        return len(self._iterable)
+
+
+class MapMemoryCache(torch.utils.data.Dataset):
+    """
+    Wrap a dataset so that, whenever a new item is returned, it is saved to memory.
+    """
+
+    def __init__(self, dataset):
+        self.dataset = dataset
+        self._cache = [None] * len(dataset)
+
+    def __getitem__(self, n):
+        if self._cache[n]:
+            return self._cache[n]
+
+        item = self.dataset[n]
+        self._cache[n] = item
+
+        return item
+
+    def __len__(self):
+        return len(self.dataset)
+
+
+class Processed(torch.utils.data.Dataset):
+
+    def __init__(self, process_datapoint, dataset):
+        self.process_datapoint = process_datapoint
+        self.dataset = dataset
+
+    def __getitem__(self, n):
+        try:
+            item = self.dataset[n]
+            return self.process_datapoint(item)
+        except (FileNotFoundError, RuntimeError):
```

Comment:
What's the purpose of catching these exception? As in, in what circumstances do you expect to see them in this particular case?

Reply:
The previous version of LibriSpeech would raise an error since a file was missing by design of the dataset. This is no longer the case. Removed.